### ISSUE TITLE
use own okHttp client for AppAuth 

### ISSUE THREAD

**Issue by theScrabi** [owncloud/android#2997]
~[AppAuthConnection](https://github.com/owncloud/android/blob/f7d720d/owncloudApp/src/main/java/com/owncloud/android/authentication/oauth/OAuthConnectionBuilder.kt) still uses ApacheHttp, which we already replaced with OkHttp in other places. This should be changed in order not to have two network libraries. At the moment having two libraries causes several issues like having to handle TLS separately or having two different user agents.~

AppAuth uses plain java [HttpUrlConnection](https://docs.oracle.com/javase/8/docs/api/java/net/HttpURLConnection.html). However in the java Runtime HttpUrlConnection is an abstract class that is implemented internally by the android build in [okHttp](https://android.googlesource.com/platform/external/okhttp/+/master/okhttp-urlconnection/src/main/java/com/squareup/okhttp/internal/huc/HttpURLConnectionImpl.java). This however is done in a way that it behaves as close as possible to the java implementation, this also means that an internal instance of the okHttp client instead of our instance is used.
In order to get our settings and interceptors to work with AppAuth we therefore have to create a HttpUrlConnection instance that uses our own okHttp client.

Tasks:

- research how to make Android HttpUrlConnection use our okHttp client

Need more research to know the impact/scope

*TL/DR:* We decided on modifying AppAuth, and add a wrapper to its network library. This way we can inject our own okHttpClient into AppAuth. The related AppAuth PR can be found [here](https://github.com/openid/AppAuth-Android/pull/621)

appauth: https://github.com/openid/AppAuth-Android/pull/621
app: https://github.com/owncloud/android/pull/3022
library: https://github.com/owncloud/android-library/pull/359

Related: https://github.com/owncloud/android/issues/3009

**Comment by michaelstingl** [comment]
@abelgardep could this help with the unloggable requests in https://github.com/owncloud/android/issues/2887 ?

**Comment by abelgardep** [comment]
Well, if we were able to use the same OkHttpClient to perform auth requests, it would help definitely. I'm not sure whether it is possible. Maybe @davigonz could clarify a little bit since he researched how to integrate AppAuth.

**Comment by theScrabi** [comment]
Ok so we are NOT using ApacheHttp. AppAuth uses plain HttpUrlConnection. Which in the android Runtime is an abstract class that gets implemented by Android builtin okHttp library. However in order to "bend" the AppAuth to do connections through our own okHttp client instance we need implement HttpUrlConnection our selfs. Just like android does it internally:
https://android.googlesource.com/platform/external/okhttp/+/master/okhttp-urlconnection/src/main/java/com/squareup/okhttp/internal/huc/HttpURLConnectionImpl.java

**Comment by theScrabi** [comment]
Ok sollution for this is rather simple:
The two libraries need to be added to the `build.gradle` of the App module:
```
    implementation "com.squareup.okhttp3:okhttp:4.6.0"
    implementation "com.squareup.okhttp3:okhttp-urlconnection:3.0.0"
```
Then after sync in the [OAuthConnectionBuilder](https://github.com/owncloud/android/blob/a61d8ff9faed77e447f6cb13d0bd30de9ce0e49b/owncloudApp/src/main/java/com/owncloud/android/authentication/oauth/OAuthConnectionBuilder.kt) we
need to replace the line
```
conn = URL(uri.toString()).openConnection() as HttpsURLConnection
```
with
```
conn = HttpURLConnectionImpl(URL(uri.toString()), HttpClient.getOkHttpClient())
```
whereas the `HttpClient.getOkHttpClient()` singleton is the client that is responsible for all of the other connections of our app. It also contains has the interceptor for the HTTP loging.
This singleton should later on be replaced with proper dependency injection of course but this is not part of this issue.

**Comment by theScrabi** [comment]
We have been looking for serveral libraries with which we could replace AppAuth if it was easyer to use okHttp with an authentication library.

## Requirements for the library
- A custom okHttpClient can be injected into the library, so that we can use the custom okHttpClient to log requests.
- The library supports oauth and open id connect, so we can use it with our ownCloud.
- The library is well maintained, so we don't have to maintain it at some point.

## Libraries we found
### [AppAuth](https://github.com/openid/AppAuth-Android)
- **Pro**
  - We already use it and know how it works.
  - Officially supported by Google and the openID Connect project.
  - Lots of users and well maintained.
- **Cons**
  - Uses `HttpURLConnection` for network requests instead of a replacable interface.
  - Injecting an okHttpClient would require to implement `HttpURLConnection` our selfs.
  - `HttpURLConnection` is a statefull object and not just an interface which makes implementation hard and error prone.

### [Okta](https://github.com/okta/okta-oidc-android)
- **Pro**
  - Uses an interface that can easily be replaced by okHttp.
  - Commercially maintained
  - Supports openID connect and OAuth
- **Cons**
  - Aparently uses proprietary okta API before doing actual openID connect. Examples without it could not yet be found.

### [Auth0](https://github.com/auth0/Auth0.Android)
- **Pro**
  - Commercialy maintained
  - uses okHttp internally
- **Cons**
  - more research required

**Comment by michaelstingl** [comment]
> * Lots of users and well maintained.

Not sure about AppAuth maintainer status:
- https://github.com/openid/AppAuth-Android/issues/444
- https://github.com/openid/AppAuth-Android/graphs/contributors?from=2020-01-01&to=2020-11-12&type=c

**Comment by theScrabi** [comment]
They could have switched the maintainer. There have been contributions after he left: https://github.com/openid/AppAuth-Android/commits/master

**Comment by theScrabi** [comment]
Here are example request logs. The first is a login with default AppAuth, the second is with okHttp based AppAuth:
[oauth_login_httpurlconnection.log](https://github.com/owncloud/android/files/5596423/oauth_login_httpurlconnection.log)
[oauth_login_okhttp.log](https://github.com/owncloud/android/files/5596424/oauth_login_okhttp.log)

**Comment by theScrabi** [comment]
Ok so here is the current situation: I changed the [AppAuth code](https://github.com/openid/AppAuth-Android/pull/621), however since it is unlikely changes will be merged soon we had the idea of implementing OIDC support our self. This already happened to iOS and Desktop, so we do have some experience in doing that.

For a short term solution I'd like to suggest to use my fork. As the longterm solution we should implement OIDC our self.

**Comment by theScrabi** [comment]
So the AppAuth solution using okHttp seems to be working except for the token request. This appears to be unstable sometimes. There are two ways to continue now. Either keep trying with AppAuth where we already do have some working code but are not able to tell which problems might keep arising, or scrape AppAuth and try to implement OIDC client ourself, which might take a while but render us with a proper solution.

**Comment by theScrabi** [comment]
We opted for a custom implementation.
